### PR TITLE
feat(ci/cd): Publish canary releases

### PR DESCRIPTION
**Pull request by roninjin10**
- publish normal releases on workflow dispatch
- publish canary release when commits hit main

### CODE REVIEW

File: .github/workflows/npm.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Releases a canary release when new commits merge to main
+# Releases an official release when workflow dispactch is
+# manually triggered
+on:
+  workflow_dispatch:
```
Comment (roninjin10) [inline .github/workflows/npm.yml:5]: You can trigger this manually in actions tab to do a real release

File: .github/workflows/npm.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  push:
+    branches:
+      - main
```
Comment (roninjin10) [inline .github/workflows/npm.yml:8]: Everytime commits are added to main branch a canary release is triggered

File: .github/workflows/npm.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      - name: Outdated files detected, run `yarn tablegen` and commit them
+        run: |
+          if [[ -n "$(git status --porcelain)" ]]; then
+            git status
+            git --no-pager diff
+            exit 1
+          fi
+
+      - name: Generate gas reports
+        run: yarn gas-report
```
Comment (roninjin10) [inline .github/workflows/npm.yml:52]: I copy pasted all this from build

File: .github/workflows/npm.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# Releases a canary release when new commits merge to main
+# Releases an official release when workflow dispactch is
+# manually triggered
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+
+name: NPM Build 📦
```
Comment (roninjin10) [inline .github/workflows/npm.yml:10]:
```suggestion
name: NPM Release 📦
```

File: .github/workflows/npm.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      - name: Generate gas reports
+        run: yarn gas-report
+
+      - name: Outdated files detected, run `yarn gas-report` and commit them
+        run: |
+          if [[ -n "$(git status --porcelain)" ]]; then
+            git status
+            git --no-pager diff
+            exit 1
+          fi
```
Comment (roninjin10) [inline .github/workflows/npm.yml:60]:
```suggestion
```
We can either add more checks here if we want to protect releases via running tests or remove stuff like this to make it so this doesn't block canary releases

File: .github/workflows/npm.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        if: github.event_name != 'workflow_dispatch'
+        run: yarn release:canary
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+      - name: Publishing new release to NPM 📦
+        if: github.event_name == 'workflow_dispatch'
+        run: yarn release
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (roninjin10) [inline .github/workflows/npm.yml:72]: I guessed this is the env variable we need to set in the github actions secrets.   I am not 100% sure lerna is expecting NPM_TOKEN

File: .github/workflows/npm.yml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      - name: Outdated files detected, run `yarn gas-report` and commit them
+        run: |
+          if [[ -n "$(git status --porcelain)" ]]; then
+            git status
+            git --no-pager diff
+            exit 1
+          fi
+
+      - name: Publishing canary releases to NPM 📦
+        if: github.event_name != 'workflow_dispatch'
```
Comment (roninjin10) [inline .github/workflows/npm.yml:63]: Only run canary on push events

File: .github/workflows/npm.yml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        run: yarn release:canary
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+      - name: Publishing new release to NPM 📦
+        if: github.event_name == 'workflow_dispatch'
```
Comment (roninjin10) [inline .github/workflows/npm.yml:69]: only do a real release on workflow-dispatch

File: package.json [hunk 9]
```diff
@@ -64,6 +64,7 @@
     "prettier": "prettier --write 'packages/**/*.ts'",
     "lint": "eslint . --ext .ts",
     "release": "yarn && lerna publish --no-private --force-publish && yarn retype:updateversion",
+    "release:canary": "yarn && lerna publish --canary --no-private --force-publish && yarn retype:updateversion",
```
Comment (roninjin10) [inline package.json:67]:
```suggestion
    "release:canary": "yarn && lerna publish --canary --no-private --force-publish",
```
I'm guessing we don't need retype:updateversion?

File: package.json [hunk 10]
```diff
@@ -64,6 +64,7 @@
     "prettier": "prettier --write 'packages/**/*.ts'",
     "lint": "eslint . --ext .ts",
     "release": "yarn && lerna publish --no-private --force-publish && yarn retype:updateversion",
```
Comment (roninjin10) [inline package.json:66]: This might need something like changesets to work correctly

**Comment by alvrs** [comment]
Moving this to #533 to simplify testing the action in the mud repo before merging

**Pull request closed without merging**